[Commerce SDK React Integration] Product List Page

bendvc opened the pull request:
# Description

This PR replaced the used of `getProps` with the _hooks_ from the `commerce-sdk-react` library.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [x] **Other changes** (non-breaking changes that does not fit any of the above)

# Changes

- (change1)

# How to Test-Drive This PR

- Preview this branch and contrast the functionality of the product list page and search page with the same pages on the previous version [here](https://pwa-kit.mobify-storefront.com/)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

wjhsf committed:
New year, new look! (#876)

* New year, new look!

* Use JS to compute current year.

* Small bump to max deps allowed.

Needed to so that non-code changes will pass CI.
Hitting the limit can be addressed later.

bendvc committed:
Initial commit of plp using commerce-sdk-react

alexvuong committed:
Update from 2.5.0 (#881)

* Starting release process for 2.5.0

* allow commerce react sdk to release

* bump version to 2.5.0 (#879)

* bump version to 2.5.0

* bump max packages

* Begin development on 2.6.0

vcua-mobify committed:
Set commerce-sdk-react package to private (#882)

bendvc committed:
Wishlist integration

vmarta committed:
test-commerce: comment out not-implemented customer hooks (#877)

bendvc committed:
Clean up refine values

bendvc committed:
Update out dated lock files

bendvc committed:
Fix tests

wjhsf commented on packages/template-retail-react-app/app/hooks/use-search-params.js:
```diff
@@ -71,10 +74,13 @@ export const parse = (searchParamsStr, parseRefine = true) => {
-    params.refine = Array.isArray(params.refine) ? params.refine : [params.refine]
+    params.refine = Array.isArray(params.refine)
+        ? params.refine
+        : [params.refine].filter((v) => Boolean(v))
```
```suggestion
        : [params.refine].filter(Boolean)
```
`Boolean` is already a method that takes a single param. 😉

wjhsf commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -84,115 +89,197 @@ const REFINEMENT_DISALLOW_LIST = ['c_isNew']
+        case 404:
+            throw new HTTPNotFound('Category Not Found.')
+        default:
+            throw new HTTPError('Unknown Error Occured.')
```
```suggestion
            throw new HTTPError(`HTTP Error ${errorStatus} occurred.`)
```
It's not unknown!

wjhsf commented on packages/template-retail-react-app/app/hooks/use-search-params.js:
```diff
@@ -42,6 +42,9 @@ export const useSearchParams = (searchParams = DEFAULT_SEARCH_PARAMS) => {
 export const stringify = (searchParamsObj) => {
     let searchParamsObjCopy = {...searchParamsObj}
 
+    // Remove our copy of the original refinement value so it's not stringified.
+    delete searchParamsObjCopy._refine
```
This is not necessary if `_refine` is a non-enumerable property (set via `Object.defineProperty(params, '_refine', { value: params.refine })`).

bendvc replied:
I like that. But as it site it won't work without some further refactoring because the `useSearchParams` hook uses destructuring which is effected by the enumerability of the object.

wjhsf commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -84,115 +89,197 @@ const REFINEMENT_DISALLOW_LIST = ['c_isNew']
+    const basePath = `${location.pathname}${location.search}`
+    const showNoResults = !isLoading && productSearchResult && !productSearchResult?.hits
+    const {total, sortingOptions} = productSearchResult || {}
+    let selectedSortingOptionLabel = productSearchResult?.sortingOptions?.find(
+        (option) => option.id === productSearchResult?.selectedSortingOption
+    )
+
+    // API does not always return a selected sorting order
+    if (!selectedSortingOptionLabel) {
+        selectedSortingOptionLabel = productSearchResult?.sortingOptions?.[0]
+    }
```
```suggestion
    const selectedSortingOptionLabel = sortingOptions?.find(
        (option) => option.id === productSearchResult?.selectedSortingOption
    ) ?? sortingOptions?.[0]
```
1. You've already destructured `sortingOptions` into a variable
2. Use `??` for the fallback value

bendvc committed:
Fix test setup

bendvc committed:
Update packages/template-retail-react-app/app/hooks/use-search-params.js

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Update packages/template-retail-react-app/app/pages/product-list/index.jsx

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Update packages/template-retail-react-app/app/pages/product-list/index.jsx

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Fix linting errors

alexvuong commented on packages/template-retail-react-app/app/utils/test-utils.js:
```diff
@@ -134,31 +140,42 @@ export const TestProviders = ({
+                    redirectURI={`${window.location.origin}/testcallback`}
+                >
+                    <MultiSiteProvider site={site} locale={locale} buildUrl={buildUrl}>
+                        <CommerceAPIProvider value={api}>
```
Nit: Can we rename this one to <_CommerceAPIProvider> to make it more distinct from the new commerceApi Provider?

alexvuong commented on packages/template-retail-react-app/app/utils/test-utils.js:
```diff
@@ -182,12 +199,32 @@ TestProviders.propTypes = {
-export const renderWithProviders = (children, options) =>
-    render(children, {
+export const renderWithProviders = (children, options) => {
+    const TestProvidersWithDataAPI = withReactQuery(withLegacyGetProps(TestProviders), {
```
The naming of this Provider is a bit confusing. it does not reflect what it really is. Maybe TestProvidersWithReactQuery?

bendvc replied:
I thought about that name too.. but we are using the data api "getProps" as well.. hence that name.

alexvuong replied:
Do we need to keep `withLegacyGetProps` at the end of the integration? If not, I think the name makes sense

bendvc replied:
I think it's wise to leave the `withLegacyGetProps` in there as to catch any kind of integration errors if some come up. 

Lets leave the name as is for now.

GoodNightBuddy committed:
Change padding (#899)

snyk-bot committed:
[Snyk] Security upgrade eslint-import-resolver-webpack from 0.10.0 to 0.13.1 (#887)

* fix: packages/internal-lib-build/package.json & packages/internal-lib-build/package-lock.json to reduce vulnerabilities

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-JS-DEBUG-3227433

* upgrade eslint-webpack-plugin-import, regen lock file

Co-authored-by: Alex Vuong <alex.vuong@salesforce.com>
Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

kevinxh committed:
Cache SLAS callback using request processor (#884)

* cache callback in request processor

* fix import path

* cache callback for a year

* use native URLSearchParams

* revert use native URLSearchParams

alexvuong committed:
Rotate fingerprint ssh for deploy commerce sdk doc (#905)

kevinxh committed:
move msw mocks to global

kevinxh committed:
fix tests

bendvc committed:
Merge remote-tracking branch 'origin/global-msw' into refactor-product-list

bendvc committed:
Fix tests, lint, fix merge issues

bendvc committed:
Increase vendor limit

bendvc committed:
Make old commerce API more obvious

bendvc committed:
Merge branch 'feature/integrate-commerce-sdk-react' into refactor-product-list

bendvc committed:
Update lock files

alexvuong commented on packages/template-typescript-minimal/package-lock.json:
```diff
@@ -5,12 +5,12 @@
 	"requires": true,
 	"dependencies": {
 		"@babel/runtime": {
-			"version": "7.20.7",
```
Is this change relevant? Same with the lock file in the test project

bendvc replied:
Because of the merge from develop, I had to fix the lock files. It was easier to just blow it away and regenerate it.

alexvuong commented on packages/commerce-sdk-react/package-lock.json:
```diff
@@ -5,9 +5,9 @@
 	"requires": true,
 	"dependencies": {
 		"@adobe/css-tools": {
-			"version": "4.0.2",
```
Why do lock files use the older version when we re-generate them?

bendvc replied:
No idea. Could be because of slight npm versions changes, could be because of caching. Hard to really say to be honest. 
I can put it back if you like, but I imaging that the lock files will be re-generate a few times before this branch is merged into develop

alexvuong commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -258,6 +349,8 @@ const ProductList = (props) => {
             {...rest}
         >
+            {/* <div>isLoading: {isLoading ? 'Loading' : 'Loaded'}</div>
+            <div>isFetching: {isFetching ? 'Fetching' : 'Fetched'}</div> */}
```
Don't forget to remove this

alexvuong commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -290,6 +383,7 @@ const ProductList = (props) => {
                             <SelectedRefinements
                                 filters={productSearchResult?.refinements}
                                 toggleFilter={toggleFilter}
+                                handleReset={() => resetFilters()}
```
```suggestion
                                handleReset={resetFilters}
```

alexvuong commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -84,115 +89,193 @@ const REFINEMENT_DISALLOW_LIST = ['c_isNew']
+    })
+
+    /**************** Query Actions ****************/
+    const {isLoading, isFetching, data: productSearchResult} = useProductSearch(
```
Can we make this loading more contextual since there are other loading state in the page? `isProductSearchLoading`.

alexvuong replied:
same with isFetching

alexvuong commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -84,115 +89,193 @@ const REFINEMENT_DISALLOW_LIST = ['c_isNew']
     }
 
-    const basePath = `${location.pathname}${location.search}`
     // Reset scroll position when `isLoaded` becomes `true`.
```
Is the comment `isLoaded` still relevant?

alexvuong commented:
Saw a type warning on one of the filters
![image](https://user-images.githubusercontent.com/52219283/213036810-81a9eb8e-78b8-405e-9dc6-31ce9d1eff23.png)

bendvc commented:
> Saw a type warning on one of the filters ![image](https://user-images.githubusercontent.com/52219283/213036810-81a9eb8e-78b8-405e-9dc6-31ce9d1eff23.png)

It looks like this is happening on develop as well. I'll see if it's an easy fix and squeeze it in this PR. If not, then we'll have to create a ticket for it.

bendvc committed:
PR feedback

bendvc committed:
Fix size filters

bendvc committed:
Correctly parse search params types

bfeister reviewed (approved):
Yeah I get it re: lots of test changes, but I think of it as "deferred maintenance" in this case. Looks good 🏅
